### PR TITLE
Custom devices new

### PR DESCRIPTION
**Pull request by elral**
This PR adds support for custom devices. This makes it a lot of easier to implement own devices.
A defined interface for the data exchange is introduced.
There was already PR #214 which was closed as major changes are implemented in this new branch.

The code for custom devices will have an own repo, for now it's https://github.com/elral/MobiFlight-CustomDevices which must be transferred to Mobiflight when this PR gets merged. Additionally the script `get_CustomDevices.py` and the workflows `ci.yml` / `release.yml` must be adapted in this case. They point for now to my repo to get the PR compiled on github.

On first executing of `pio run [-e environment]` the repo for custom devices will be cloned into the folder `./CustomDevices`. On all next executing of `pio run [-e xyz]` this folder gets updated and all available custom devices gets compiled.

All the required code for a custom device must be in one folder. How to setup this code is explained in https://github.com/elral/MobiFlight-Connector/wiki/How-to-set-up-a-custom-device (which must also be transferred to the Mobiflight WiKi) and in a template under `./CustomDevices/_template/`.
To get it automatically compiled during a `pio run` the `platformio.ini` must be adapted with one additional line which points to an .ini file whithin the device folder.
Some examples are already included.

For the naming convention the names of the standard environments are changed. The `mobiflight` is now part of this environment and not anymore added by the `get_version.py` script. This enables to have firmware file names starting not only with mobiflight but also with a different name (most likely [company]name of the user/company).

The boards.h files are added for using custom devices. The config.cpp is added to read a config with custom devices. The commandmessenger.cpp is added with new functions to call the custom device functions.  A new folder `/src/CustomDevices` is added where the  basic functions to support custom devices are included. These files are always the same and must not be changed for a new custom device.

### CODE REVIEW

**Comment by elral** [comment]
@neilenns @GioCC may I ask you to have also a look on this PR?

@neilenns would be great to check espacially the changes in the automatic build process: `ci.yml `/ `release.yml` / `get_version.py`.

@GioCC I used the code from your PR for the TM1637 implementation, but deleted all MAX7219 related code to have another example for a custom device. If this is not OK for you, I will delete it or change to another lib. Or you try to setup this example as a custom device ;)

File: .github/workflows/release.yml
Comment (elral) [inline .github/workflows/release.yml:59]: copy all .hex and .uf2 files to the zip folder, not only these with mobiflight*

File: get_CustomDevices.py
Comment (elral) [inline get_CustomDevices.py:12]: This file gets executed on each build from an environment. It clones the repo for custom devices or checks for updates

File: get_version.py
Comment (elral) [inline get_version.py:26]: As now firmware names should start with the supporter/user name, `mobiflight` is moved to the environments

File: platformio.ini
Comment (elral) [inline platformio.ini:27]: Include builts from external files which are part of the custom device repo and cloned.
For new custom devices onyl one entry has to be added. No more changes on this repo are necessary.

File: platformio.ini
Comment (elral) [inline platformio.ini:163]: for all boards the environments are added with `mobiflight` to get the existent firmware file name (see above)

File: src/CommandMessenger.cpp
Comment (elral) [inline src/CommandMessenger.cpp:95]: new callback for custom devices

File: src/Config.cpp
Comment (elral) [inline src/Config.cpp:443]: new case for supporting custom devices. All other changes should be reformatting

File: src/MF_CustomDevice/CustomDevice.cpp
Comment (elral) [inline src/MF_CustomDevice/CustomDevice.cpp:86]: New functions for all custom devices.
No changes for new custom devices are required as all must use these functions.

File: src/mobiflight.cpp
Comment (elral) [inline src/mobiflight.cpp:214]: Defines within the platformio.ini for custom devices can enable these update() functions. As this is defined in the customer platformio.ini no changes are requires for new ones

**Review by elral — COMMENTED** [review]
Some more explanations of the code changes

**Comment by GioCC** [comment]
WOW, @ElRal, that's a really impressive work! Congratulations are in order!
It'll take some time to properly go through the whole set of modifications, but I see the initial documentation is already well made.
Regarding the TM1637 code, needless to say, I don't see a reason why I wouldn't be OK with that; of course I still prefer for it to be included natively, but until then it can perfectly serve its purpose as an example.

**Comment by GioCC** [comment]
One small thing I noticed: in [all_custom_devices_mega.board.json](https://github.com/elral/MobiFlight-CustomDevices/blob/main/_all_CustomDevices/connector_files/all_custom_devices_mega.board.json) in the other repo:
```
    "CustomDeviceType": [
      "MOBIFLIGHT_TEMPLATE",  // Does the template need to be included?
      "MOBIFLIGHT_TEMPLATE",  // Should probably be MOBIFLIGHT_GENERICI2C
      "KAV_EFIS",
      "KAV_FCU",
      "MOBIFLIGHT_GNC255",
      "4TM1637",  // Should probably be MOBIFLIGHT_4TM1637
      "6TM1637"  // Should probably be MOBIFLIGHT_6TM1637
    ]
```

**Comment by elral** [comment]
> WOW, @elral, that's a really impressive work! Congratulations are in order!

Thanks a lot, took me all the last weeks.

> Regarding the TM1637 code, needless to say, I don't see a reason why I wouldn't be OK with that; of course I still prefer for it to be included natively, but until then it can perfectly serve its purpose as an example.

Great, and I see it also to have a native support.

> One small thing I noticed: in [all_custom_devices_mega.board.json](https://github.com/elral/MobiFlight-CustomDevices/blob/main/_all_CustomDevices/connector_files/all_custom_devices_mega.board.json) in the other repo:

    "CustomDeviceType": [
      "MOBIFLIGHT_TEMPLATE",  // Does the template need to be included?
      "MOBIFLIGHT_TEMPLATE",  // Should probably be MOBIFLIGHT_GENERICI2C
      "KAV_EFIS",
      "KAV_FCU",
      "MOBIFLIGHT_GNC255",
      "4TM1637",  // Should probably be MOBIFLIGHT_4TM1637
      "6TM1637"  // Should probably be MOBIFLIGHT_6TM1637
    ]

Perfect, great finding. Will change it. The code within MFCustomDevice.cpp isn't optimized, just copied from the other files. I have done it in this way to show what has to be done for setting it up.
I expect at least some more naming issues (not reall good in this) as it was a lot of files working on.

Two times `MOBIFLIGHT_TEMPLATE` was intended, just to show how two different device classes should be handled.
So yes, I also think that the template should not be included. Just left it in to see how everything is setup and to discuss it (at lest the template makes exactly nothing :) ) 
Oh, and it seems that I forget to implement the generic I2C devide. I will chech this.

Thanks very much!

**Comment by GioCC** [comment]
I have written a few notes for myself while starting to look at the examples in the CustomDevices repository; I though I'd attach it here in case it contains something useful to add to the documentation (or not).
Please let me know if you notice anything wrong... I'll look further into the matter as soon as I'm able.
[MF_CustomDevices_notes.md](https://github.com/MobiFlight/MobiFlight-FirmwareSource/files/12671781/MF_CustomDevices_notes.md)

**Comment by elral** [comment]
Oh, I forgot to mention that espacially the TM1637 is good example to have the so called configParameter. This one is already implemented in the firmware, but mot in the connector. The number of digits could be set into this configParameter. This would avoid the use of two device.json files and having only one device type in the code. Additionally the decimal point could be set, so no additional message to set this would be required.

**Comment by elral** [comment]
> I have written a few notes for myself while starting to look at the examples in the CustomDevices repository; I though I'd attach it here in case it contains something useful to add to the documentation (or not). Please let me know if you notice anything wrong... 

Perfect! It's exactly what the intention is, but much better described (than I could do ;) ) I am impressed how fast you got everything.

> I'll look further into the matter as soon as I'm able. 

This would be great.

The only thing I have to think about is where to copy your comments exactly. Another way would be to transfer the description soon to the Mobiflight WiKi. Then you could directly add your improvements there.

File: get_CustomDevices.py
Comment (neilenns) [reply get_CustomDevices.py:12]: Why is this script needed? The repo already gets cloned by the changes made to the GitHub workflow scripts.

File: .github/workflows/ci.yml
Comment (neilenns) [inline .github/workflows/ci.yml:41]: Since we'll never have custom devices in default MF firmware should we have this commented out in the repo? Then tell people building firmware with a custom device where to go to uncomment it for their own builds?

**Review by neilenns — COMMENTED** [review]
Asked a couple of clarification questions

File: .github/workflows/ci.yml
Comment (elral) [reply .github/workflows/ci.yml:41]: Why should we not provide "some" custom devices which are officially reviewed and released?
And how do get users a firmware file if they uncomment this with the correct FW Version? I am doing it always by changing the get_version.py script ;)

Comment (GioCC) [reply .github/workflows/ci.yml:41]: One question about this. The script has Platform.IO run to to build the firmware. Doesn't Platform.IO itself clone the CustomDevices repo? What's the reason why this also has to be done in the script? Asking because I'm not that experienced in this.

File: get_CustomDevices.py
Comment (GioCC) [reply get_CustomDevices.py:12]: OK, turns out I asked the same question (in a complementary way) before reading this... 
The way I see it, I'm inclined to say that the cloning here should remain - for local builds it has to be there somewhere, For automated builds it can be in the script if this instance isn't applicable to that case for whatever reason (which is my question elsewhere).

**Review by neilenns — COMMENTED** [review]
Forgot to ask a question

File: get_CustomDevices.py
Comment (neilenns) [reply get_CustomDevices.py:12]: Also as it stands now this is causing an error in the automated build process:

```
git --work-tree=/home/runner/work/MobiFlight-FirmwareSource/MobiFlight-FirmwareSource\CustomDevices --git-dir=/home/runner/work/MobiFlight-FirmwareSource/MobiFlight-FirmwareSource\CustomDevices\.git pull origin main --depth 100
fatal: not a git repository: '/home/runner/work/MobiFlight-FirmwareSource/MobiFlight-FirmwareSourceCustomDevices.git'
```

File: .github/workflows/ci.yml
Comment (neilenns) [reply .github/workflows/ci.yml:41]: Until we have actual custom devices that ship with Mobiflight firmware directly we shouldn't be running a pull for them in the build.

When the day comes that we do decide to ship custom devices with the core firmware that PR can enable the pulling of the custom devices repo.

File: src/Config.cpp
Comment (neilenns) [inline src/Config.cpp:255]: These are all causing unused variable warnings in the build:

```
src/Config.cpp:250:14: warning: unused variable 'length' [-Wunused-variable]
     uint16_t length    = MFeeprom.get_length();
              ^~~~~~
src/Config.cpp:251:14: warning: unused variable 'temp' [-Wunused-variable]
     char     temp      = 0;
              ^~~~
src/Config.cpp:252:14: warning: unused variable 'adrPin' [-Wunused-variable]
     uint16_t adrPin    = 0;
              ^~~~~~
src/Config.cpp:253:14: warning: unused variable 'adrType' [-Wunused-variable]
     uint16_t adrType   = 0;
              ^~~~~~~
src/Config.cpp:254:14: warning: unused variable 'adrConfig' [-Wunused-variable]
     uint16_t adrConfig = 0;
```

File: platformio.ini
Comment (neilenns) [inline platformio.ini:20]: The template shouldn't be included, correct?

File: platformio.ini
Comment (neilenns) [reply platformio.ini:27]: Are any of these actually going to ship with the main mobiflight firmware? If so, where is the build step that clones them?

File: _Boards/Atmel/Board_Mega/MFBoards.h
Comment (neilenns) [inline _Boards/Atmel/Board_Mega/MFBoards.h:76]: We screwed up when adding shifters because we did output first and just called them "shifter". Then when I came and added input shifters they got called "inputshifter".

Let's not make the same mistake again with custom devices and get it right the first time. Call them CUSTOM_OUTPUT_DEVICES and make sure all the relevant files are called "CustomOutput" device.

File: .github/workflows/ci.yml
Comment (neilenns) [reply .github/workflows/ci.yml:41]: Ohhh never mind, I see there's custom devices coming already just the repo needs to be moved over. Right?

File: _Boards/Atmel/Board_Mega/MFBoards.h
Comment (GioCC) [reply _Boards/Atmel/Board_Mega/MFBoards.h:76]: I second that, good point.

File: .github/workflows/ci.yml
Comment (elral) [reply .github/workflows/ci.yml:41]: Yes

Comment (elral) [reply .github/workflows/ci.yml:41]: Once my repo is transferred the links have also to be changed

Comment (elral) [reply .github/workflows/ci.yml:41]: > One question about this. The script has Platform.IO run to to build the firmware. Doesn't Platform.IO itself clone the CustomDevices repo? What's the reason why this also has to be done in the script? Asking because I'm not that experienced in this.

Difficult for me to explain.
If you have a plain Installation, the folder /CustomDevice does not exist. Therefore the files listed in platformio.ini are not found and these environments get not into the list of environments to be compiled. On the next executing of the platformio.ini these files are available on they are getting into the list of environments to be compiled. As the automatic build process has always not the CustomDevice folder and pio run is only executed once, the custom device repo has to be cloned before.
Or what just comes in mind, a pio run - e mega and next a pio run could be done

File: src/Config.cpp
Comment (neilenns) [reply src/Config.cpp:271]: Ok phew!

File: platformio.ini
Comment (elral) [reply platformio.ini:20]: Yeah, will remove it. Also in the all devices environment.

File: src/Config.cpp
Comment (elral) [reply src/Config.cpp:255]: I have to check it later

Comment (elral) [reply src/Config.cpp]: Can't remember what I did ;)
Changed it.

File: get_CustomDevices.py
Comment (elral) [reply get_CustomDevices.py]: This was due to using backslashes in the path.

File: platformio.ini
Comment (elral) [reply platformio.ini]: Done!

File: .github/workflows/ci.yml
Comment (neilenns) [reply .github/workflows/ci.yml:41]: I can't remember if I asked this yet, but are these custom devices from the repo that has to move over intended to ship as part of the core Mobiflight firmware? They're just added using the custom device method instead of being coded as new specific devices?

Comment (elral) [reply .github/workflows/ci.yml:41]: These are fully functional devices. From my side all can be moved to the new Mobiflight repo. And from my side all could be shipped with the main firmware.
But in the end it is the decision of @DocMoebiuz which one will be moved/shipped.

Comment (neilenns) [reply .github/workflows/ci.yml:41]: @DocMoebiuz If the intention is to ship these as part of the core firmware they should be checked in to the firmware repo, not some external repo and then built in. If the intention isn't to ship them then these build steps should be disabled in the official firmware repo with guidance given to 3rd parties about how to enable them for their own builds.

Aside from complicating the build and local development process we lose valuable git benefits by storing them somewhere else:

1. There's no easy way to know which iteration of the devices were built into a version of the firmware. The commit hash won't tell us and the source won't be included in the .zip attached to each release.
2. Any changes made to those devices won't be included in the automatically generated release notes. We'll need to remember to manually add them every time.
3. It'll be yet another repo for us to have issue tracking. We'll have to remember to look over there for issues related to code that directly ships in the firmware.

File: _Boards/Atmel/Board_Mega/MFBoards.h
Comment (elral) [reply _Boards/Atmel/Board_Mega/MFBoards.h:76]: Needed some time to think about.

I am still not convinced that it's required to rename it (in contrast to the shifters). The user can setup a custom device as input device, output device or both (how to name in this case). In the end it is a custom devices and the code defines what it is. From the firmware side mainly only a handler is missing which gets called if a defined action is ececuted. For the connector it's function (input and/or output) could be marked in the accordingly device.json file.

File: .github/workflows/ci.yml
Comment (DocMoebiuz) [reply .github/workflows/ci.yml:41]: > @DocMoebiuz If the intention is to ship these as part of the core firmware they should be checked in to the firmware repo, not some external repo and then built in. If the intention isn't to ship them then these build steps should be disabled in the official firmware repo with guidance given to 3rd parties about how to enable them for their own builds.
> 
> Aside from complicating the build and local development process we lose valuable git benefits by storing them somewhere else:
> 
> 1. There's no easy way to know which iteration of the devices were built into a version of the firmware. The commit hash won't tell us and the source won't be included in the .zip attached to each release.
> 2. Any changes made to those devices won't be included in the automatically generated release notes. We'll need to remember to manually add them every time.
> 3. It'll be yet another repo for us to have issue tracking. We'll have to remember to look over there for issues related to code that directly ships in the firmware.

@neilenns thanks for your valuable feedback! In general, I want to create an opportunity for the community to extend MobiFlight at their on pace and without having me (or us) become  the bottleneck. I don't want to be part of the custom device life cycle. At the moment I don't see any custom device become built-in in the official mobiflight firmware. I envision and hope that people come up with their custom device implementations and at the moment I expect them to compile a specific firmware versions, hence board type, for it - or leave it up to the individual user to compile a custom firmware on their own.

@elral has experimented with including different custom devices as repository references. It reminds me of npm or pip where you install a library. There are still some lines of code that have to be adapted to use those custom devices and expose them correctly. But i think for the first release that is perfectly fine and makes it very easy to get started. our assumption is that a user who develops custom firmware will find the process easy enough.

So to specifically answer your questions:
> 1. There's no easy way to know which iteration of the devices were built into a version of the firmware. The commit hash won't tell us and the source won't be included in the .zip attached to each release.
... that's OK. the custom firmware can have it's own version. It will have a base firmware version of the official mobiflight firmware and a combination of custom device versions... but it is the responsiblity of the custom firmware creator to maintain this up-to-date in case there are improved versions available.

> 2. Any changes made to those devices won't be included in the automatically generated release notes. We'll need to remember to manually add them every time.
... that's OK, I see these devices outside of the official releases anyways. If we have an official MobiFlight Community Repository with curated custom devices (like npm registry) then we can have release notes in that repository.

> 3. It'll be yet another repo for us to have issue tracking. We'll have to remember to look over there for issues related to code that directly ships in the firmware.
... that's OK, this is the responsibility of the respective device developer or repository maintainer. For the beforementioned official MobiFlight Custom Device repository I am sure that we can find a feasible solution.

And to say it one more time: I see a big advantage for the community to decouple the generic devices from the official MobiFlight Core Firmware. We keep it this way on purpose so that the community can move on its own pace. We can still think of some incubator process later which might help to bring popular and stable custom devices to the official MobiFlight Community Repository.

Maybe we can find a cool way to automatically create custom firmwares in the future, where the user has the ability to select from a list of available devices the ones they like to have in their firmware. I am convinced that this first iteration provides already enough options to unlock a lot of creativity.

**Review by DocMoebiuz — APPROVED** [review]
Thanks! It is time to merge this.

**Pull request merged**
